File: compiler/bench/strlen.c
// expected: 25500000
// Length of 1M strings
// Checksum: sum of all lengths

#include <stdio.h>

char buf[64];

void fill_buf(int n) {
    int len = (n % 50) + 1;
    for (int i = 0; i < len; i++) {
        buf[i] = 'A';
    }
    buf[len] = '\0';
}

int my_strlen(void) {
    int len = 0;
    while (buf[len]) len++;
    return len;
}

int main(void) {
    long sum = 0;
    for (int i = 0; i < 1000000; i++) {
        fill_buf(i);
        sum += my_strlen();
    }
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/strlen.fs
\ expected: 25500000
\ Length of 1M strings
\ Checksum: sum of all lengths

create buf 64 allot

: fill-buf ( n -- ) dup 50 mod 1+ 0 do 65 buf i + c! loop 50 mod 1+ buf + 0 swap c! ;
: my-strlen ( -- len ) 0 begin buf over + c@ while 1+ repeat ;

: bench-strlen ( -- sum )
  0
  1000000 0 do
    i fill-buf
    my-strlen +
  loop ;

: main bench-strlen . cr ;
